### ISSUE TITLE
Evaluating a model requires knowledge of the training set

### ISSUE THREAD

**Issue by InakiRaba91** [open-mmlab/mmsegmentation#3471]
Thanks for your error report and we appreciate it a lot.

**Describe the bug**
I'm running into an issue when evaluatinga model on a test set, i.e.: calling [runner.test()](https://github.com/open-mmlab/mmsegmentation/blob/main/tools/test.py#L119). If visualization is active, `SegVisualizationHook` tries to access the runner.iter [property](https://github.com/open-mmlab/mmsegmentation/blob/main/mmseg/engine/hooks/visualization_hook.py#L98). This in turn tries to eventually [instantiate](https://github.com/open-mmlab/mmengine/blob/main/mmengine/runner/runner.py#L586-L592) the train_dataloader.

The only way to avoid that is making sure [every training arg in the config is None](https://github.com/open-mmlab/mmengine/blob/main/mmengine/runner/runner.py#L307-L316). I think the code should be able to infer that in test mode, there's no need whatsoever to instantiate the training dataset. 

**Bug fix**

It seems MMDet was able to handle this case by having [mode-specific methods](https://github.com/open-mmlab/mmdetection/blob/main/mmdet/engine/hooks/visualization_hook.py#L109), as opposed to a general one, as is the case in [MMSeg](https://github.com/open-mmlab/mmsegmentation/blob/main/mmseg/engine/hooks/visualization_hook.py#L66). Would it be alright to extend this implementation to this repo as well? I'm more than happy to submit a PR, just wanted to double check first.
